flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,696 @@
+  private final Table table;
+  private final boolean overwrite;
+  private final boolean upsertMode;
+  private final DistributionMode distributionMode;
+  private final List<String> equalityFieldColumns;
+  private final List<Integer> equalityFieldIds;
+  private final FileFormat dataFileFormat;
+  private final int workerPoolSize;
+  private final long targetDataFileSize;
+  private final String uidPrefix;
+  private final Map<String, String> snapshotProperties;
+  private final RowType flinkRowType;
+  private Committer<FilesCommittable> committer;
+
+  private IcebergSink(
+      TableLoader tableLoader,
+      @Nullable Table table,
+      List<Integer> equalityFieldIds,
+      List<String> equalityFieldColumns,
+      @Nullable String uidPrefix,
+      Map<String, String> snapshotProperties,
+      boolean upsertMode,
+      boolean overwrite,
+      TableSchema tableSchema,
+      DistributionMode distributionMode,
+      int workerPoolSize,
+      long targetDataFileSize,
+      FileFormat dataFileFormat) {
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
+    Preconditions.checkNotNull(table, "Table shouldn't be null");
+
+    this.tableLoader = tableLoader;
+    this.table = table;
+    this.equalityFieldIds = equalityFieldIds;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+    this.flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+    this.upsertMode = upsertMode;
+    this.overwrite = overwrite;
+    this.distributionMode = distributionMode;
+    this.workerPoolSize = workerPoolSize;
+    this.targetDataFileSize = targetDataFileSize;
+    this.dataFileFormat = dataFileFormat;
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(
+        inputDataStream,
+        equalityFieldIds,
+        table.spec(),
+        table.schema(),
+        flinkRowType,
+        distributionMode,
+        equalityFieldColumns);
+  }
+
+  @Override
+  public StreamWriter createWriter(InitContext context) {
+    return restoreWriter(context, null);
+  }
+
+  @Override
+  public StreamWriter restoreWriter(
+      InitContext context, Collection<StreamWriterState> recoveredState) {
+    RowDataTaskWriterFactory taskWriterFactory =
+        new RowDataTaskWriterFactory(
+            SerializableTable.copyOf(table),
+            flinkRowType,
+            targetDataFileSize,
+            dataFileFormat,
+            equalityFieldIds,
+            upsertMode);
+    StreamWriter streamWriter =
+        new StreamWriter(
+            table.name(),
+            taskWriterFactory,
+            context.getSubtaskId(),
+            context.getNumberOfParallelSubtasks());
+    if (recoveredState == null) {
+      return streamWriter;
+    }
+
+    return streamWriter.restoreWriter(recoveredState);
```

Comment:
I found the `StreamWriter#restoreWrite` method name a little confusing. `initializedState` seems more accurate to me.

Reply:
Nit: I would go with `initializeState` (without the d) but I agree with @stevenzwu . It matches the argument name, and in the case of using an empty collection to avoid the `if` block to return early it also reads a bit more fluently to me.

Reply:
Nice. Thx.